### PR TITLE
[Integration] Expose length-aware batching in all ModelHandler subclasses

### PR DESCRIPTION
**Pull request by Eliaaazzz**
## Summary

Addresses #37531.

This PR completes the smart bucketing integration for Python `RunInference` by exposing `batch_length_fn` and `batch_bucket_boundaries` on all concrete `ModelHandler` implementations.

The underlying batching support already exists in the base layer. The missing piece was that many user-facing handlers did not surface these options, which made length-aware batching effectively unavailable for a large part of the inference API surface. With this change, users can enable smart bucketing directly from the handler constructor across supported backends.

## What Changed

This change adds `batch_length_fn` and `batch_bucket_boundaries` to 16 concrete handlers across the following backends:

- PyTorch
- HuggingFace
- scikit-learn
- TensorFlow
- ONNX
- XGBoost
- TensorRT
- vLLM
- Vertex AI
- Gemini

Implementation details:
- Handlers that inherit from `ModelHandler` now pass the new parameters through to `super().__init__()`
- Remote handlers that manage batching kwargs directly (`GeminiModelHandler` and `VertexAIModelHandlerJSON`) now wire the values into `_batching_kwargs`

## Testing

Added test coverage in `base_test.py` for both behavior and wiring:

- an end-to-end `RunInferenceLengthAwareBatchingTest` that verifies short and long string inputs are bucketed into separate batches under `FnApiRunner`
- a `HandlerBucketingKwargsForwardingTest` that checks each concrete handler forwards `batch_length_fn` and `batch_bucket_boundaries` into `batch_elements_kwargs()`
- follow-up fixes to keep the forwarding tests hermetic, especially for HuggingFace pipeline validation and Vertex AI endpoint liveness checks

## Context

This is the final integration piece for smart bucketing:

- Part 1: #37532
- Part 2: #37565

Together, these changes make length-aware batching usable through the public Python inference handlers rather than only at the base implementation layer.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

### CODE REVIEW

**Comment by Eliaaazzz** [comment]
./gemini review

File: sdks/python/apache_beam/ml/inference/base_test.py [hunk 1]
```diff
@@ -2319,6 +2319,234 @@ def test_batching_kwargs_none_values_omitted(self):
...
+    examples = ['a', 'cccccc', 'bb', 'ddddddd']
+    with TestPipeline('FnApiRunner') as p:
+      results = (
+          p
+          | beam.Create(examples, reshuffle=False)
+          | base.RunInference(handler))
+      assert_that(results, equal_to(['a:2', 'bb:2', 'cccccc:7', 'ddddddd:7']))
+
+
+class HandlerBucketingKwargsForwardingTest(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (Eliaaazzz) [reply sdks/python/apache_beam/ml/inference/base_test.py]: Good point. These cases all exercise the same forwarding assertion with different handler constructors, so I can consolidate them into a single data-driven test using `subTest`.

**Comment by Eliaaazzz** [comment]
@shunping Hi could you please have a look? All three failed checks show the same infrastructure error: "The self-hosted runner lost communication with the server." No tests actually failed before the runner disconnected. This appears to be a transient Apache Beam CI infrastructure issue with the self-hosted runner dropping offline during execution, and it is unrelated to the changes in this PR.

**Comment by shunping** [comment]
@damccorm, could you help with this?

File: sdks/python/apache_beam/ml/inference/base_test.py [hunk 2]
```diff
@@ -2319,6 +2322,264 @@ def test_batching_kwargs_none_values_omitted(self):
...
+  def _assert_bucketing_kwargs_forwarded(self, handler):
+    kwargs = handler.batch_elements_kwargs()
+    self.assertIs(kwargs['length_fn'], len)
+    self.assertEqual(kwargs['bucket_boundaries'], [32])
+
+  def _load_handler_class(self, case):
+    try:
+      module = importlib.import_module(case['module_name'])
+    except ImportError:
+      raise unittest.SkipTest(case['skip_message'])
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py]: Rather than taking in a skip_message, we can just grab the module name; that way we don't mask the cause if our skip message is wrong.

```
except ImportError as e:
     raise unittest.SkipTest(f"Could not import {e.name}")
```

File: sdks/python/apache_beam/ml/inference/base_test.py [hunk 3]
```diff
@@ -2319,6 +2322,264 @@ def test_batching_kwargs_none_values_omitted(self):
...
+    examples = ['a', 'cccccc', 'bb', 'ddddddd']
+    with TestPipeline('FnApiRunner') as p:
+      results = (
+          p
+          | beam.Create(examples, reshuffle=False)
+          | base.RunInference(handler))
+      assert_that(results, equal_to(['a:2', 'bb:2', 'cccccc:7', 'ddddddd:7']))
+
+
+class HandlerBucketingKwargsForwardingTest(unittest.TestCase):
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py]: Actually, I would probably just remove these tests actually; as it stands, the only thing they protect us against is removing one of these arguments, which almost certainly would need to be intentional or obvious from the diff.

File: sdks/python/apache_beam/ml/inference/base_test.py (hunk 2 above)
Comment (Eliaaazzz) [reply sdks/python/apache_beam/ml/inference/base_test.py]: > Rather than taking in a skip_message, we can just grab the module name; that way we don't mask the cause if our skip message is wrong.
> 
> ```
> except ImportError as e:
>      raise unittest.SkipTest(f"Could not import {e.name}")
> ```

Thanks, that makes sense. I’ll derive the skip reason directly from the `ImportError` rather than passing a separate `skip_message`.

File: sdks/python/apache_beam/ml/inference/base_test.py (hunk 3 above)
Comment (Eliaaazzz) [reply sdks/python/apache_beam/ml/inference/base_test.py]: > Actually, I would probably just remove these tests actually; as it stands, the only thing they protect us against is removing one of these arguments, which almost certainly would need to be intentional or obvious from the diff.

And agreed on the kwargs forwarding tests as well. I’ll remove them since they don’t add much protection beyond catching obvious or intentional argument changes.

**Comment by Eliaaazzz** [comment]
@damccorm Thanks for the review, I removed the kwargs forwarding tests as suggested. Since the separate `skip_message` handling was only used in that test class, removing it also addressed the import-skip comment there.

**Review by damccorm — APPROVED** [review]
Thanks, this LGTM. I'll merge after checks complete
